compact.go

```diff
@@ -757,9 +757,19 @@ func (c *LeveledCompactor) populateBlock(blocks []BlockReader, meta *BlockMeta,
 		}
 
 		for i, chk := range chks {
-			if chk.MinTime < meta.MinTime || chk.MaxTime > meta.MaxTime {
-				return errors.Errorf("found chunk with minTime: %d maxTime: %d outside of compacted minTime: %d maxTime: %d",
-					chk.MinTime, chk.MaxTime, meta.MinTime, meta.MaxTime)
+			// Re-encode head chunks that are still open (being appended to) or
+			// outside the compacted MaxTime range.
+			// The chunk.Bytes() method is not safe for open chunks hence the re-encoding.
+			// This happens when snapshotting the head block.
+			//
+			// Block time range is half-open: [meta.MinTime, meta.MaxTime) and
+			// chunks are closed hence the chk.MaxTime >= meta.MaxTime check.
+			if _, isHeadChunk := chk.Chunk.(*safeChunk); isHeadChunk && chk.MaxTime >= meta.MaxTime {
```

Comment:
Checking for safeChunk is a bit magic, I'd explicitly pass in a bool to the function.

Reply:
+1, also the bool would avoid checking every chunk for a head chunk.

Reply:
Can agree. but I would rather have separate method for `populateFromHead` instead of bool

Reply:
I am thinking the same as `populateBlock` is already getting too messy and too long.

Reply:
so we are waiting on this right?

Also `safeChunk` is not that bad I guess with given comment IMO - same as `http Handler` and casting `http.Flusher` to flush.

Reply:
yeah I can spend a bit more time, but so far can't see a good clean way to separate `populateBlock` and `populateHead` without duplicating a lot of code.

Reply:
https://github.com/prometheus/tsdb/pull/641#issuecomment-507986776